File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/consent/scope.rb
# frozen_string_literal: true

require_relative "scopes"
require_relative "errors"

module IAIso
  module Consent
    # A verified consent scope.
    class Scope
      attr_reader :token, :subject, :scopes, :execution_id, :jti,
                  :issued_at, :expires_at, :metadata

      def initialize(token:, subject:, scopes:, execution_id:, jti:,
                     issued_at:, expires_at:, metadata: {})
        @token = token
        @subject = subject
        @scopes = scopes.freeze
        @execution_id = execution_id
        @jti = jti
        @issued_at = issued_at.to_i
        @expires_at = expires_at.to_i
        @metadata = metadata || {}
        freeze
      end

      def grants?(requested)
        Scopes.granted(@scopes, requested)
      end

      # Raises InsufficientScopeError if any of `required` is not granted.
      def require_scopes!(*required)
        required.flatten.each do |r|
          unless grants?(r)
            raise InsufficientScopeError.new(r, @scopes)
          end
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-php/src/Consent/Scope.php
<?php

declare(strict_types=1);

namespace IAIso\Consent;

/** A verified consent scope. */
final class Scope
{
    public function __construct(
        public readonly string $token,
        public readonly string $subject,
        /** @var string[] */
        public readonly array $scopes,
        public readonly ?string $executionId,
        public readonly string $jti,
        public readonly int $issuedAt,
        public readonly int $expiresAt,
        /** @var array<string,mixed> */
        public readonly array $metadata = [],
    ) {
    }

    /** True iff the verified scope set grants {@code $requested}. */
    public function grants(string $requested): bool
    {
        return Scopes::granted($this->scopes, $requested);
    }

    /** Throws {@see InsufficientScopeException} if any of {@code $required} is not granted. */
    public function requireScopes(array $required): void
    {
        foreach ($required as $r) {
            if (!$this->grants($r)) {
                throw new InsufficientScopeException(
                    "scope '$r' not granted; have [" . implode(', ', $this->scopes) . ']');
            }
        }
    }
}
